docs/src/markdown.ts

```diff
@@ -0,0 +1,266 @@
+import fs from "fs/promises";
+import path from "path";
+import { unified } from "unified";
+import remarkParse from "remark-parse";
+import remarkRehype from "remark-rehype";
+import rehypeRaw from "rehype-raw";
+import { visit } from "unist-util-visit";
+import GitHubSlugger from "github-slugger";
+import matter from "gray-matter";
+
+export interface Document {
+  /** the Markdown file itself, without from-matter */
+  content: string;
+
+  /** the path to this markdown file */
+  path: string;
+
+  /** the headings found in this markdown file */
+  headings: string[];
+
+  frontMatter: {
+    title: string;
+    description: string;
+  };
+}
+
+export type ErrorType = "link" | "hash" | "source" | "related";
+
+export type LinkError = {
+  type: ErrorType;
+  href: string;
+  doc: Document;
+};
+
+export const DOCS_PATH = ".";
+export const EXCLUDED_HASHES = ["top"];
+
+const slugger = new GitHubSlugger();
+
+/** Collect the paths of all .mdx files in the passed directories */
+const getAllMdxFilePaths = async (
+  directoriesToScan: string[],
+  fileList: string[] = []
+): Promise<string[]> => {
+  for (const dir of directoriesToScan) {
+    const dirPath = path.join(".", dir);
+    const files = await fs.readdir(dirPath);
+    for (const file of files) {
+      const filePath = path.join(dirPath, file);
+      const stats = await fs.stat(filePath);
+      if (stats.isDirectory()) {
+        fileList = await getAllMdxFilePaths([filePath], fileList);
+      } else if (path.extname(file) === ".mdx") {
+        fileList.push(filePath);
+      }
+    }
+  }
+
+  return fileList;
+};
+
+// Returns the slugs of all headings in a tree
+const getHeadingsFromMarkdownTree = (
+  tree: ReturnType<typeof markdownProcessor.parse>
+): string[] => {
+  const headings: string[] = [];
+  slugger.reset();
+
+  visit(tree, "heading", (node) => {
+    let headingText = "";
+    // Account for headings with inline code blocks by concatenating the
+    // text values of all children of a heading node.
+    visit(node, (innerNode: any) => {
+      if (innerNode.value) {
+        headingText += innerNode.value;
+      }
+    });
+    const slugified = slugger.slug(headingText);
+    headings.push(slugified);
+  });
+
+  return headings;
+};
+
+/** Create a processor to parse MDX content */
+const markdownProcessor = unified()
+  .use(remarkParse)
+  .use(remarkRehype)
+  .use(rehypeRaw)
+  .use(function compiler() {
+    // A compiler is required, and we only need the AST, so we can
+    // just return it.
+    // @ts-ignore
+    this.Compiler = function treeCompiler(tree) {
+      return tree;
+    };
+  });
+
+const normalizePath = (filePath: string): string => {
```

Comment:
Total nit, but this is more of converting a file path to a url path, than normalizing a path.